revzilla/payeezy
# Refactor code to fix Dialyzer issues

## Changes
As the changelog says:
* Elixir has been updated to version 1.11.4.
* Test helpers have been moved from `lib` to `test/support` directory.
* Dialyxir has been added to dependencies.
* Processing API response has been refactored to avoid overriding typespecs defined in HTTPoison.

## QA
1. Pull this branch.
2. Run `mix deps.get`.
3. Run tests: `mix test`. Confirm that they all pass.
4. Run `mix dialyzer`. Confirm that it doesn't report any issues.
5. Open `config/config.exs` and addd there the following configuration (no worries about the credentials, they are taken from the public Payeezy sandbox, available [here](https://developer.payeezy.com/payeezy-api/apis/post/transactions-8)):
```elixir
config :payeezy,
  apikey: "y6pWAJNyJyjGv66IsVuWnklkKUPFbb0a",
  token: "fdoa-a480ce8951daa73262734cf102641994c1e55e7cdf4c02b6",
  apisecret: "86fbae7030253af3cd15faef2a1f4b67353e41fb6799f576b5093ae52901e6f7",
  endpoint: "https://api-cert.payeezy.com/v1/transactions"
```
6. Run `iex -S mix`. Create the following variables there:
```elixir
params = %{
  "amount" => "400",
  "transaction_type" => "purchase",
  "method" => "valuelink",
  "currency_code" => "USD",
  "post_date" => "04032016",
  "reference_3" => "svspin",
  "valuelink" => %{
    "cardholder_name" => "Joe Smith",
    "cc_number" => "7777045839985463",
    "credit_card_type" => "Gift"
  }
}

params2 = %{
  "amount" => "400",
  "transaction_type" => "purchasee",
  "method" => "valuelink",
  "currency_code" => "USD",
  "post_date" => "04032016",
  "reference_3" => "svspin",
  "valuelink" => %{
    "cardholder_name" => "Joe Smith",
    "cc_number" => "7777045839985463",
    "credit_card_type" => "Gift"
  }
}
```
7. Still in `IEx` console, run `Payeezy.Client.post("", params)`. Confirm that the result makes sense and looks good. You should get something like that:
```elixir
{:ok,
 %{
   "amount" => "400",
   "bank_message" => "Insufficient fund",
   "bank_resp_code" => "521",
   "correlation_id" => "228.5181369677748",
   "currency" => "USD",
   "gateway_message" => "Transaction Normal",
   "gateway_resp_code" => "00",
   "local_currency_code" => "840",
   "method" => "valuelink",
   "post_date" => "04032016",
   "retrieval_ref_no" => "220505",
   "transaction_status" => "declined",
   "transaction_tag" => "4724503596",
   "transaction_type" => "purchase",
   "validation_status" => "success",
   "value_link_auth_code" => "521",
   "valuelink" => %{
     "cardholder_name" => "Joe Smith",
     "cc_number" => "############5463",
     "credit_card_type" => "Gift",
     "current_balance" => "2.59",
     "previous_balance" => "2.59"
   }
 }}
```
8. Now run `Payeezy.Client.post("", params2)`. Confirm that the result makes sense and looks good. You should get something like that:
```elixir
{:error,
 %{
   "Error" => %{
     "messages" => [
       %{
         "code" => "invalid_transaction_type",
         "description" => "The transaction type is not supported"
       }
     ]
   },
   "amount" => "400",
   "correlation_id" => "228.5181376184436",
   "currency" => "USD",
   "method" => "valuelink",
   "transaction_status" => "Not Processed",
   "transaction_type" => "purchasee",
   "validation_status" => "failed"
 }}
```
9. Now run `Payeezy.Client.post("a/b/c", params)`. You should get `{:error, :not_found}` as the result.
10. Change `endpoint` in `config.exs` to `127.0.0.1:3000`. Restart `IEx` console.
11. Run `Payeezy.Client.post("", %{})`. Confirm that it doesn't throw any errors, you should get the following tuple as the result: `{:error, ":econnrefused"}`.
12. That's all folks!
